vault/plugin_catalog.go

```diff
@@ -832,7 +832,15 @@ func (c *PluginCatalog) get(ctx context.Context, name string, pluginType consts.
 	}
 
 	builtinVersion := versions.GetBuiltinVersion(pluginType, name)
-	if version == "" || version == builtinVersion {
+	if version == builtinVersion {
+		// Don't return the builtin if it's shadowed by an unversioned plugin.
+		unversioned, err := c.get(ctx, name, pluginType, "")
+		if err == nil && unversioned != nil && !unversioned.Builtin {
```

Comment:
Might we worth factoring out a common function for this check. Seems to be duplicated in 3 places.

Reply:
I agree there's opportunity for improvement here, but after noodling about with it for a while, due to the differences in context between the 3 places I think the refactoring is a bit more major than I'm comfortable doing for a fix that I want to backport.